### PR TITLE
Sentiment analysis implmenttion in pure Keras.

### PR DESCRIPTION
**Pull request by sculd**
- This is an update for the sentiment analysis model's pure Keras version.
-- Converting it from the version using Tensorflow's estimator, as it has a issue that affects the accuracy of the model negatively.
- The implementation is with the reference to paddle version.
-- Adjustment of the hyperparameters was done to achieve the accuracy of ~90%

### CODE REVIEW

**Comment by sculd** [comment]
ping

File: research/sentiment_analysis/data/imdb.py [hunk 1]
```diff
@@ -1,46 +1,47 @@
...
+    A tuple of length 4, for training and evaluation data,
+    each being an numpy array.
   """
   (x_train, y_train), (x_test, y_test) = tf.keras.datasets.imdb.load_data(
       path="imdb.npz",
       num_words=vocabulary_size,
       skip_top=0,
       maxlen=None,
-      seed=113,
+      seed=112,
```
Comment (karmel) [inline research/sentiment_analysis/data/imdb.py]: Why did this change?

File: research/sentiment_analysis/data/imdb.py [hunk 2]
```diff
@@ -1,46 +1,47 @@
...
       start_char=START_CHAR,
       oov_char=OOV_CHAR,
-      index_from=OOV_CHAR + 1)
+      index_from=OOV_CHAR+1)
```
Comment (karmel) [inline research/sentiment_analysis/data/imdb.py]: nit: spaces around `+`

File: research/sentiment_analysis/sentiment_main.py [hunk 3]
```diff
@@ -1,160 +1,79 @@
-"""The main module for sentiment analysis.
-
-The model makes use of concatenation of two CNN layers
-with different kernel sizes.
-See `sentiment_model.py` for more details about the models.
-"""
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_main.py]: Why remove the module docstring?

File: research/sentiment_analysis/sentiment_main.py [hunk 4]
```diff
@@ -1,160 +1,79 @@
...
+                              hid_dim, dataset.get_num_class(dataset_name),
+                              _DROPOUT_RATE)
+  model.summary()
 
-  return estimator
+  model.compile(loss="categorical_crossentropy",
+                optimizer="rmsprop",
+                metrics=["accuracy"])
 
+  print("loading the data")
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_main.py]: Prefer tf.logging

**Comment by karmel** [comment]
Thanks-- and really happy to hear that the accuracy delta is gone.

File: research/sentiment_analysis/data/imdb.py (hunk 2 above)
Comment (sculd) [reply research/sentiment_analysis/data/imdb.py]: ack

File: research/sentiment_analysis/data/dataset.py [hunk 5]
```diff
@@ -38,9 +34,9 @@ def get_num_class(dataset):
...
     ValueError: if the dataset value is not valid.
   Returns:
-    str: The dataset name.
+    int: The number of label classes.
   """
   if dataset == DATASET_IMDB:
     return imdb.NUM_CLASS
   else:
-    raise ValueError("unsupported dataset: " + dataset)
+    raise ValueError("unsupported dataset: " + dataset)
```
[the remark this answers is not in the document]
Comment (sculd) [reply research/sentiment_analysis/data/dataset.py]: ack

File: research/sentiment_analysis/sentiment_main.py (hunk 4 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_main.py]: ack

File: research/sentiment_analysis/data/dataset.py [hunk 6]
```diff
@@ -3,28 +3,24 @@
...
     vocabulary_size: The number of the most frequent tokens
       to be used from the corpus.
     sentence_length: The number of words in each sentence.
       Longer sentences get cut, shorter ones padded.
-    repeat: The number of epoch.
   Raises:
     ValueError: if the dataset value is not valid.
   Returns:
-    A tuple of training and evaluation input function.
+    A tuple of length 4, for training and evaluation data,
```
[the remark this answers is not in the document]
Comment (sculd) [reply research/sentiment_analysis/data/dataset.py]: ack

File: research/sentiment_analysis/data/imdb.py (hunk 1 above)
Comment (sculd) [reply research/sentiment_analysis/data/imdb.py]: It was likely to be changed while I was adjusting the hyper parameter. I am redoing the run to see the same accuracy is achieved with 113 seed.

File: research/sentiment_analysis/sentiment_main.py (hunk 3 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_main.py]: I am adding doc string back to the main module also adding to other modules.

File: research/sentiment_analysis/data/imdb.py (hunk 1 above)
Comment (sculd) [reply research/sentiment_analysis/data/imdb.py]: I verified that the run with the seed of 113 gives 90.5% accuracy.

**Comment by sculd** [comment]
ping

File: research/sentiment_analysis/data/dataset.py [hunk 7]
```diff
@@ -16,7 +21,8 @@ def load(dataset, vocabulary_size, sentence_length):
   Raises:
     ValueError: if the dataset value is not valid.
   Returns:
-    A tuple of length 4, for training and evaluation data,
+    A tuple of length 4, for training sentences, labels,
+    evaluation sentences, nd evaluation labels,
```
Comment (karmel) [inline research/sentiment_analysis/data/dataset.py]: nit: and

File: research/sentiment_analysis/data/imdb.py [hunk 8]
```diff
@@ -1,3 +1,6 @@
+"""IMDB Dataset module for sentiment analysis.
+"""
```
Comment (karmel) [inline research/sentiment_analysis/data/imdb.py]: nit: You can remove the extra \n in the docstring

File: research/sentiment_analysis/data/imdb.py (hunk 2 above)
Comment (karmel) [reply research/sentiment_analysis/data/imdb.py]: Please fix.

File: research/sentiment_analysis/sentiment_main.py [hunk 9]
```diff
@@ -1,6 +1,14 @@
...
+
+The model makes use of concatenation of two CNN layers with
+different kernel sizes. See `sentiment_model.py`
+for more details about the models.
+"""
+
 import argparse
 from data import dataset
 import sentiment_model
+import tensorflow as tf
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_main.py]: Please fix import order-- should be python builtins, newline, thirdparty (including tf), newline, local lib.

File: research/sentiment_analysis/sentiment_main.py [hunk 10]
```diff
@@ -30,15 +38,15 @@ def run_model(dataset_name, emb_dim, voc_size, sen_len,
                 optimizer="rmsprop",
                 metrics=["accuracy"])
 
-  print("loading the data")
+  tf.logging.info("loading the data")
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_main.py]: nit: capitalize Loading

File: research/sentiment_analysis/sentiment_main.py [hunk 11]
```diff
@@ -30,15 +38,15 @@ def run_model(dataset_name, emb_dim, voc_size, sen_len,
...
   x_train, y_train, x_test, y_test = dataset.load(
       dataset_name, voc_size, sen_len)
 
   model.fit(x_train, y_train, batch_size=batch_size,
             validation_split=0.4, epochs=epochs)
   score = model.evaluate(x_test, y_test, batch_size=batch_size)
-  print("score")
-  print(score)
+  tf.logging.info("score")
+  tf.logging.info(score)
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_main.py]: nit: Score: ### would be more clear and concise in the logs.

File: research/sentiment_analysis/sentiment_model.py [hunk 12]
```diff
@@ -1,3 +1,9 @@
+"""Model for the sentiment analysis.
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_model.py]: nit: no `the`

File: research/sentiment_analysis/sentiment_model.py [hunk 13]
```diff
@@ -25,9 +31,9 @@ def __init__(self, emb_dim, num_words, sentence_length, hid_dim,
       tf.keras.models.Model: A Keras model.
     """
 
-    layer = tf.keras.layers.Input(shape=(sentence_length,), dtype=tf.int32)
+    sen = tf.keras.layers.Input(shape=(sentence_length,), dtype=tf.int32)
```
Comment (karmel) [inline research/sentiment_analysis/sentiment_model.py]: What does sen stand for here? Can you use a clearer var name?

File: research/sentiment_analysis/data/dataset.py (hunk 7 above)
Comment (sculd) [reply research/sentiment_analysis/data/dataset.py]: ack

File: research/sentiment_analysis/data/imdb.py (hunk 8 above)
Comment (sculd) [reply research/sentiment_analysis/data/imdb.py]: ack

File: research/sentiment_analysis/sentiment_main.py (hunk 10 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_main.py]: ack

File: research/sentiment_analysis/sentiment_main.py (hunk 11 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_main.py]: ack

File: research/sentiment_analysis/sentiment_model.py (hunk 12 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_model.py]: ack

File: research/sentiment_analysis/sentiment_model.py (hunk 13 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_model.py]: ack

File: research/sentiment_analysis/sentiment_main.py (hunk 9 above)
Comment (sculd) [reply research/sentiment_analysis/sentiment_main.py]: ack

**Comment by sculd** [comment]
ping

File: research/sentiment_analysis/sentiment_main.py (hunk 11 above)
Comment (karmel) [reply research/sentiment_analysis/sentiment_main.py]: Sorry, I meant-- tf.logging.info("Score: {}".format(score)) or similar.

**Comment by sculd** [comment]
ping (for the final comment).
Thank you for reviewing the PR. After the checks are complete, can someone merge the PR please?

**Comment by karmel** [comment]
Thanks!

**Pull request merged**